bulk REST requests

peterthomassen opened the pull request:
Finally!

A few notes:

- I tried to make this a few separate commits, to avoid confusion. The main commit may still seem large (but it's a large feature, too). I can try to split it up a bit more if it is too clumsy, but the changes within it are a bit intertwined, so I'm not sure if it is really possible.
- There are one or two unrelated commits that I didn't bother to make a separate PR for, as they are not urgent.

- [x] One of the e2e tests currently fails due to a minor problem; I'll try to fix it during the next days. I don't think that needs to hold back review.
- [x] e2e tests for RRset deletion via bulk requests (setting `'records': []`) are missing; I'll add them over the next days.
- [x] e2e tests checking DNS output are mostly disabled (status "pending", with a dash) because the DNS library doesn't really give us what we want (not even TTLs, except for A/AAAA). We should try to find another DNS library.

Have fun, and 1000 thanks for reviewing! <3

peterthomassen commented:
I added e2e tests for RRset deletion via bulk requests (setting `'records': []`).

peterthomassen commented:
A few details about input validation for bulk requests: There are three stages during which exceptions can occur:

1. `serializer.is_valid()`: Those are errors like missing JSON fields, negative TTL and such (based on `RRset` model metadata).
2. Uniqueness. This is tested in the run-up of the database write operation, in `RRsetSerializer._save()`. Here, we also track that one bulk request does not contain two orders for the same `(subname, type)`.
3. Lastly, the type or record contents may be deemed invalid by pdns. This only occurs after the tentative database write (before committing transaction).

Errors are collected in each stage and presented as a list of errors, with each list item referring to one part of the bulk request, in the same order. Parts that did not cause errors have `{}`, and parts with errors have a `dict` describing the error. Only in step 3 it is not possible to distinguish by bulk part, as the pdns response does not contain information on which RRset was faulty.

The treatment of stages 1 and 2 may mean that one bulk part appears without an error in stage 1 (`{}`) if another part has an error. Only after this error is resolved, the request will reach stage 2, at which point an error may occur for the bulk part that previously seemed valid.

While it would be possible to merge stages 1 and 2, it would rip apart the usual structure of DRF serializers, and I'm also not sure if it is worth the effort. It probably suffices to mention this behavior in the docs.

Errors in stages 1 and 2 cause status code `400` (regardless of the exact reason(s) which may vary across bulk parts), and errors from stage 3 cause status code `422`.

peterthomassen commented:
Outstanding issues:

- [ ] test bulk-messing with SOA and other restricted types
- [ ] `PATCH`ing `[{"subname": "d.1", "ttl": 50, "type": "AAA"}]` does not behave correctly (weird type)
- [ ] `PATCH` generally does not require the `ttl` field, but how about new RRsets with `PATCH`? And at least for delection (`records = []`), `ttl` should not be required.
- [ ] docs

Not aware of anything else.

peterthomassen commented:
All done except docs, which I prefer to write once the bulk API has been reviewed.

@nils-wisiol Can you look at this PR when you find some time? Thanks!

nils-wisiol commented:
Before we merge this, I'd like to add further e2e tests. To that end I updated the e2e dns test interface to use a different dns library, so that dns responses include more details (most importantly, the TTL). Also, it's quite easy to add further RR set types.

[Example usage](c6da51e8c4932f1f50394d8c5b08778c5a7bf90d) of the [new code](b7ce50c5836bd4b93b2ed261c298646925c97cd5):

    describe("a domain endpoint", function () {

        var domain = 'e2etest-' + require("uuid").v4() + '.dedyn.io';

        before(function () {
            return expect(chakram.post('/domains/', {'name': domain})).to.have.status(201);
        });

        function postAndPropagationTest(subname, domain, type, records, ttl) {
            var queryName = subname ? subname + '.' + domain : domain;
            return describe('posting ' + type + ' records at ' + queryName, function () {
                before(function () {
                    return expect(chakram.post(
                        '/domains/' + domain + '/rrsets/',
                        {
                            'subname': subname,
                            'type': type,
                            'records': records,
                            'ttl': ttl,
                        }
                    )).to.have.status(201);
                });
                it('can be set', function(){});
                it('TTL propagates to pdns', function () {
                    return expect(chakram.resolve(queryName, type)).to.have.dns.ttl(ttl);
                });
                it('records propagate to pdns', function () {
                    return expect(chakram.resolveStr(queryName, type)).to.have.members(records);
                });
                it('records propagate to pdns (reverse order)', function () {
                    return expect(chakram.resolveStr(queryName, type)).to.have.members(records.reverse());
                });
            });
        }

        postAndPropagationTest('', domain, 'A', ['1.1.1.1', '2.2.2.2'], 42);
        postAndPropagationTest('a', domain, 'A', ['7.7.7.7', '8.8.8.8'], 41);
        postAndPropagationTest('', domain, 'AAAA', ['::1'], 77);
        postAndPropagationTest('a', domain, 'AAAA', ['bade::affe', 'dead::beef'], 71);
        postAndPropagationTest('', domain, 'MX', ['10 example.net.', '20 a.example.net.'], 60);
        // Note that this is not yet working as the TLSA decoding is missing from dns-packet.
        // postAndPropagationTest('', domain, 'TLSA',
        //     ['3 1 1 1306c288896771534a5b60744a535e1025f79b0b64e1bbc89278ba4d413cb9f1'], 120);

    });

More RR set types [can easily be added](https://github.com/desec-io/desec-stack/blob/b7ce50c5836bd4b93b2ed261c298646925c97cd5/test/e2e/setup.js#L139-L144).

Note that to my best knowledge, `.to.have.members()` is [order-insensitive](c6da51e8c4932f1f50394d8c5b08778c5a7bf90d) (contrary to what you said). We should verify your claim before merging 6d97421.

I propose you cherry-pick commits d6e8371485a9432d060ac8f3c392f7c29693fe5e and b7ce50c5836bd4b93b2ed261c298646925c97cd5 into your PR and adopt your e2e tests before we proceed. Note that the latter breaks existing test cases that rely on `resolve()`, as the API changed.

peterthomassen commented on api/desecapi/settings.py:
```diff
@@ -181,7 +181,7 @@
     EMAIL_BACKEND = 'django.core.mail.backends.dummy.EmailBackend'
     ABUSE_BY_REMOTE_IP_LIMIT = 100
-    ABUSE_BY_REMOTE_IP_PERIOD_HRS = 4800
+    ABUSE_BY_REMOTE_IP_PERIOD_HRS = 2
```
Yes, but do we want to disable the limit? We may want to have the functionality available.

peterthomassen replied:
Ok. I'll also add a pending test for user locking.

peterthomassen commented on api/desecapi/settings.py:
```diff
@@ -181,7 +181,7 @@
+    ABUSE_BY_REMOTE_IP_PERIOD_HRS = 2
     ABUSE_BY_EMAIL_HOSTNAME_LIMIT = 100
-    ABUSE_BY_EMAIL_HOSTNAME_PERIOD_HRS = 2400
+    ABUSE_BY_EMAIL_HOSTNAME_PERIOD_HRS = 1
```
Yes, but do we want to disable the limit? We may want to have the functionality available.

peterthomassen commented on test/e2e/spec/api_spec.js:
```diff
@@ -127,46 +151,813 @@ describe("API", function () {
                     return expect(chakram.post('/domains/', {'name': domain})).to.have.status(201);
                 });
 
+                it("has correct JSON schema");
```
No. That's a "pending test" (neither failing nor passing); shows up with "-" instead of "x" or checkmark in the results. It's meant to say that this test has no implementation yet.

peterthomassen commented:
Thanks for doing all that work! :+1: 

> Note that to my best knowledge, .to.have.members() is order-insensitive (contrary to what you said). We should verify your claim before merging 6d97421.

I said that `.members()` is /supposed/ to be order-insensitive, but in my experience with it, it was not, for unknown reasons. I'll be very happy if it turns out to actually work as advertised.

> I propose you cherry-pick commits d6e8371 and b7ce50c into your PR and adopt your e2e tests before we proceed. Note that the latter breaks existing test cases that rely on resolve(), as the API changed. I did not fix them as your PR moves them around and fixing it on master would just result in a big merge mess.

Sounds good.

peterthomassen commented:
I cherry-picked as you proposed and refactored all tests to use the new DNS tool. I also added implementations for all tests that were pending (i.e. they had a name only).

> Note that to my best knowledge, .to.have.members() is order-insensitive (contrary to what you said). We should verify your claim before merging 6d97421.

I managed to fix things up so that this commit can go away (already dropped it in this branch). I am not sure how I messed up previously.

So, I think, we're pretty much done! I have left the new changes as individual commits so that you can easily review them, but before merging, we may want to squash some of them. Here's a list:

1. 1519604 fix(api): correct abuse limits for test scenario
1. 91fc62a fix(api): fix hostname-based abuse check
1. 9959686 refactor(api): add method to convert dict-like RRsets to RRset objects
1. e60ebd8 feat(api): have write_rrsets() return all RRsets value (idempotence)
1. 848b38d feat(api): bulk REST requests, closes #83
1. cfb8740 fix(tests): adjust status codes when changing immutable fields
1. 6d65311 feat(dev): disable pdns cache for e2e tests
1. 2529177 fix(e2e): rewrote dns test interface
1. 0cbe0af fix(e2e): remove unnecessary flags from DNS resolve code
1. bfdd77b fix(e2e): close socket after tests so that mocha will exit
1. 33038c5 fix(e2e): allow NXDOMAIN DNS response code
1. 03f013f fix(e2e): add TXT record serialization to DNS toolset
1. 88d4c76 fix(e2e): augment TTL test method
1. 9dba795 feat(e2e): add test routine for checking RRsets in the DNS
1. 3d04245 refactor(e2e): adapt dynDNS tests to use new DNS tool
1. a3e0c33 refactor(e2e): move API tests to api_spec.js
1. 3630405 feat(tests): bulk API e2e tests
1. 011b34d refactor(e2e): adapt API tests to use new DNS tool, add pending tests
1. b0f7b55 fix(e2e): correctly do record comparisons
1. 79f1ce5 fix(api): improve validation of donation input data
1. 734d7a6 fix(api): when PATCH'ing, correctly inform about missing type field
1. 593803f fix(api): dummy TTL for RRsets up for deletion
1. f15b20e fix(api): disallow tinkering with OPT RRset
1. d146089 fix(api): disallow generic type format like TYPExxx
1. 5b9b6d2 fix(api): fix deadlock by rephrasing WHERE clause
1. c9a4f5b refactor(api): simplify and clarify in Domain.write_rrsets()

I think it will be reasonable to squash 9 -- 13, 14--15, 17--19.

Also, as `write_rrsets()` has undergone a bunch of changes, it is unclear why two of them should be their own commits, so I'd suggest to squatch 4, 22, 26.

Let me know what you think, both about the status of the PR as well as the squashing!

nils-wisiol commented:
yeah, please squash as proposed. Otherwise I'm afraid I'll get caught up in commenting on outdated changes :-)

peterthomassen commented:
done!

peterthomassen commented on api/desecapi/models.py:
```diff
@@ -7,6 +7,7 @@
 import datetime, uuid
```
If you send a `PATCH` request with empty record contents `[]`, the RRset is supposed to be deleted. This works by sending a similar requests with an empty `records` field to pdns. However, pdns requires the `ttl` field to be set, but our `PATCH` request reasonably does not require that. In this case, we set a dummy value.

(If you leave our the `records` field from our `PATCH` request, the RRset is not deleted; you can use such requests to change the TTL only.)

peterthomassen commented on test/e2e/spec/api_spec.js:
```diff
@@ -127,46 +152,837 @@ describe("API", function () {
+                        before(function () {
+                            var response = chakram.delete('/domains/' + domain + '/rrsets/delete-test.../A/');
+                            expect(response).to.have.status(204);
+                            return chakram.wait();
```
shall do.

peterthomassen commented on api/desecapi/views.py:
```diff
@@ -195,30 +199,27 @@ def create(self, request, *args, **kwargs):
-            raise ex
+        except ValidationError as e:
+            if isinstance(e.detail, dict):
+                all = e.detail.get('__all__')
```
shall fix

peterthomassen commented:
> I think commits 95ca54d and 48cbfdc should be together in one commit.

Really? They have nothing to do with each other. Why do you think so?

> The e2e tests 5fa6878 have lots of duplicate code. I know that's hardly avoidable, but let's try to reduce it just a little: maybe some of the it("propagates to the API", function () { can be refactored in a similar manner like you did with itShowsUpInPdnsAs?

Done, see  c13f49d743d364d00c48949079e39e9487f1c521. I added this as a new commit for ease of review, but before merging, we should squash it into the main bulk e2e test commit.

I also created another commit that adds testing of the JSON structure of responses of the `domain/` endpoint. Same goes for squashing before merge.

> Also, I'm a fan of using e2e tests as API documentation, too. In that case, it may be necessary to re-order the tests. What's a good structure for such documentation? It could be organized by endpoint, then by method

I agree that that would be nice to have, but I think we should open an issue for this instead of further holding up this PR.

> In any event, some documentation for this bulk features is still missing from this PR.

Yes. As mentioned in one of the earlier comments, I postponed writing it until it is clear (after your review) that the interface will not change much. (Otherwise, the whole bulk docs would have had to be rewritten.)

I'm working on this now, and I'd like to merge it in a separate PR. Are you ok with that?

> Other than a couple of small notes, I did not discover significant problems with this, and I'd like to see it in master soon.

Nice & thanks! I added replies to your code comments fixed some minor issues you pointed out. There are no outstanding issues from these comments.

Pls let me know about any show-stoppers that I may have forgotten. Thanks!

peterthomassen commented:
Just added the docs. A rendered version will be available for a few days at https://a4a.de/.u/1524251474_index.html

peterthomassen commented:
@nils-wisiol Here's a question regarding the docs. So far, our examples all have HTTPie commands because that tool doesn't require the user to know how to construct JSON objects. For the new bulk requests, however, HTTPie doesn't have a useful interface (see https://github.com/jakubroztocil/httpie/issues/121), so I added examples with curl.
Our supporter @gerhard-tinned proposed long ago (in an email) that we should always show curl examples, and maybe now is the time to do that. Do you think we should get rid of the HTTPIe examples, then?

Ideally, it would be nice to add tabs to the examples to switch between the styles, but I don't want to spend the time to set this up. (Right now, we simply use `rst2html5.py` which is very straightforward).

gerhard-tinned commented:
@peterthomassen For me, the example section of an api documentation is one of the most important part to start. It usually allows you an very easy start into whatever you try to do. 

As curl is kind of a standard for testing and checking web services, my opinion is - curl is the must have in an api documentation. 

If you add more options, ... even better. I would not remove the httpie commands. But I understand that maintaining both might be a bit much.

nils-wisiol commented:
I extracted outstanding issues into

- Add curl examples to docs #99
- add e2e tests to docs #100
- Clarify docs: all domain names need to end with a dot `.` #101

> I'm working on this now, and I'd like to merge it in a separate PR. Are you ok with that?

is there more to come?

Anyhow, I'm ready to approve this. lgtm! :)

peterthomassen commented:
Wow. Thanks for the review!

> > I'm working on this now, and I'd like to merge it in a separate PR. Are you ok with that?
> 
> is there more to come?

Not for the moment.

> don't forget the squash the last two commits!

Shall do, thanks for the reminder!
